services/src/main/java/org/exoplatform/webconferencing/WebConferencingService.java

```diff
@@ -0,0 +1,1235 @@
+    CallInfo call = readCallById(id);
+    if (call != null) {
+      call.setState(CallState.STARTED);
+      if (call.getOwner().isGroup()) {
+        String userId = currentUserId();
+        for (UserInfo part : call.getParticipants()) {
+          if (part.getType() == UserInfo.TYPE_NAME) {
+            if (userId.equals(part.getId())) {
+              part.setState(UserState.JOINED);
+            } else {
+              // it's eXo user: fire user listener for started call, but not to the starter (current user)
+              fireUserCallState(part.getId(),
+                                id,
+                                call.getProviderType(),
+                                CallState.STARTED,
+                                call.getOwner().getId(),
+                                call.getOwner().getType());
+            }
+          }
+        }
+      }
+      saveCall(call);
+    }
+    return call;
+  }
+
+  /**
+   * If call started, then notify all its parties that given participant joined. If stopped, then
+   * {@link InvalidCallStateException} will be thrown.
+   *
+   * @param id the id
+   * @return the call info or <code>null</code> if call not found
+   * @throws Exception the exception
+   */
+  public CallInfo joinCall(String id, String userId) throws Exception {
+    // TODO exception if user not a participant?
+    CallInfo call = readCallById(id);
+    if (call != null) {
+      if (CallState.STARTED.equals(call.getState())) {
+        // if (info.getOwner().isGroup()) {
+        for (UserInfo part : call.getParticipants()) {
+          if (part.getType() == UserInfo.TYPE_NAME) {
+            if (userId.equals(part.getId())) {
+              part.setState(UserState.JOINED);
+              saveCall(call);
+            }
+            // Fire user joined to all parts, including the user itself
+            fireUserCallJoined(id,
+                               call.getProviderType(),
+                               call.getOwner().getId(),
+                               call.getOwner().getType(),
+                               userId,
+                               part.getId());
+          }
+        }
+        // }
+      } else {
+        startCall(id);
+      }
+    }
+    return call;
+  }
+
+  /**
+   * If call started, then notify all its parties that given participant leaved. If stopped, then
+   * {@link InvalidCallStateException} will be thrown.
+   *
+   *
+   * @param id the id
+   * @return the call info or <code>null</code> if call not found
+   * @throws Exception the exception
+   */
+  public CallInfo leaveCall(String id, String userId) throws Exception {
+    // TODO exception if user not a participant?
+    CallInfo call = readCallById(id);
+    if (call != null) {
+      if (CallState.STARTED.equals(call.getState())) {
+        int leaved = 0;
+        boolean userLeaved = false;
+        for (UserInfo part : call.getParticipants()) {
+          if (part.getType() == UserInfo.TYPE_NAME) {
```

Comment:
I don't think we should use == with strings. I understand it is done for performances purposes, thanks to the use of intern(), but it can be error-prone since for example in the current case we must remember that getType() must always return an interned String. And if it very easy to forget this, especially if someone does some changes in few months.It could lead to incorrect results. So the performance gain does not balance the potential issues IMO.

Reply:
In general it's reasonable note. Yes, it was done with performance in mind. It's why UserInfo.getType() returns TYPE_NAME constant - thus it's safe for all cases. Indeed, even if we'll add String.equals() it will do references check first, thus still will work good for performance. Let's change the code with equals() call.